content/2017/04/11/usernames.rst

```diff
@@ -0,0 +1,138 @@
+public: yes
+author: david
+headline:
+  - tagline: '(None of them is the username)'
+tags: [
+  'Data Modeling',
+  Django,
+  Code
+  ]
+image:
+  - src: 'blog/2017/usernames/nametag.jpg'
+summary: |
+  The term “username” is ambiguous.
+  When designing a user model there are several
+  kinds of names that are useful to include.
+
+
+Three Names to Include in Your User Model
+=========================================
+
+It’s time to start a new project.
+You want users, so the first thing you do is
+add a user model with username and password fields and —
+whoa, hold up.
+
+What’s that username again?
+Is it the user’s email address?
+Is it an arbitrary nickname that the user can choose?
+Is it an automatically generated identifier that should never change?
+
+A Rose by Any Other Name Would Not Be Able To Log In
+----------------------------------------------------
+
+It’s not for nothing that they say naming things is
+`one of the hard problems in computer science
+<https://martinfowler.com/bliki/TwoHardThings.html>`_.
+For modeling user accounts there are three different name fields
+that you probably want,
+and for the sake of avoiding confusion
+let’s not call *any* of them the username:
+
+1. The **user id**.
+   This is the permanent, unchanging identifier of a particular user.
+   It may be generated randomly or it may be an incrementing number,
+   but the important thing is that it should not have any inherent meaning,
+   and it should never change.
+   That makes it safe for use internally throughout the system
+   as a foreign key to refer to the user.
+
+2. The **display name** or screen name.
+   This is what will actually be shown when referring to the user
+   in the UI and notifications (to the user themself or to other users).
+   This should be chosen by the user with few limitations
+   (except perhaps on length or to avoid excessive whitespace).
+   That gives users full flexibility to use their offline name
+   (including in other writing systems), nickname, stage name, or even emoji.
+   It should be unique among all users to avoid confusion,
```

Comment:
Yeah, I'll add some nuance. I think uniqueness should be enforced if it's going to be used more as a made-up screen name than a real name and users of the system interact with each other (too easy to impersonate someone otherwise).